README.md
# dfcleaner

> dataframe cleaning package

This package contains helper methods to clean pandas dataframe quickly and therefore simplifying the data cleaning process

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

![](assets/rsz_dfcleaner_logo.png)

## Installation

OS X & Linux:

```sh
pip3 install dfcleaner
```

Windows:

```sh
pip install dfcleaner
```

## Usage example

```Python
import pandas as pd
from dfcleaner import cleaner

cleaner.ENABLE_LOGGING = True
cleaner.LOG_DIR = './logs'

df = pd.read_csv('some_filename.csv')

df.columns = cleaner.sanitize(df.columns)
conversion_dict = cleaner.suggest_conversion_dict(df)
df = cleaner.preprocess(df,
                        column_dtype_conversion_dictionary = conversion_dict,
                        std_coeff = 1.5,
                        fill_na_method = 'median',
                        label_col = None)
```

## Development setup

```sh
pip3 install -r requirements.txt
```

Windows:

```sh
pip install -r requirements.txt
```

## Meta

M. Zahash – zahash.z@gmail.com

Distributed under the MIT license. See `LICENSE` for more information.

[https://github.com/zahash/](https://github.com/zahash/)

## Contributing

1. Fork it (<https://github.com/zahash/dfcleaner/fork>)
2. Create your feature branch (`git checkout -b feature/fooBar`)
3. Commit your changes (`git commit -am 'Add some fooBar'`)
4. Push to the branch (`git push origin feature/fooBar`)
5. Create a new Pull Request


CHANGELOG.md
# Changelog

## 1.1.1

### Added

- added logo in readme
- added an assets folder and stored png logo, svg logo

## 1.1.0

### Added

- added logging functionality to appropriate cleaner functions so that all the changes made to the dataframe by the corresponding functions will be logged inside a csv file in the specified LOG_DIR path

## 1.0.1

### Changed

- some major doc string additions and modifications in all the functions; Make the doc string look good and informative
- set default parameter of std_coeff to 1.5 in remove_outliers() function
- changed the word 'convertion' to 'conversion'

### Added

- add parameter 'label_col' to fill_nan() function

## 1.0.0

### Changed

- changed the function sanitize_column_names() to sanitize() which now takes just a array of strings and 'sanitizes' them; Introduced Breaking Changes
- rewrote tests in a cleaner way
- renamed the function suggest_col_drop() to spot_irrelevant_columns()

### Fixed

- bug fixed in the change_dtypes() function; bug fixes in the suggest_col_drop() function => identifying the 'id' columns like customer id, id, ... but NOT columns like rfid

## 0.2.1

### Fixed

- fixed multiple underscore issue and add support for
  mulitple consecutive capital letters (DNSServer -> dns_server) in CamelCase to snake_case conversion

## 0.2.0

### Added

- add default parameters to preprocess function
- add parameter 'label_col' in preprocess() to remove
  rows where the label_col is null
- add parameter 'label_col' in remove_outliers() so that
  the label col will not be considered as a column to remove outliers from
- add suggest_col_drop() to suggest which columns to drop (irrelevant columns) based on the column name; like ID, first_name, surname, ...

### Changed

- sanitize_column_names() now converts CamelCase to snake_case
- suggest_conversion_dict() now suggests 'category' dtype to numeric columns that are actually categorical; like 'has_credit_card' that has 1 or 0 as values

## 0.1.1

### Changed

- sanitize_column_names now keeps any already present underscores

## 0.1.0

### Added

- add MIT license badge in readme
- add function to suggest conversion dict

## 0.0.1

- work in progress


tests/test_cleaner.py
# run the below command from the root directory of project
# python3 -m unittest tests.test_cleaner

import unittest
import pandas as pd
import numpy as np
from dfcleaner.cleaner import sanitize, change_dtypes, remove_outliers, fill_nan, preprocess, suggest_conversion_dict, spot_irrelevant_columns


class TestDataCleaner(unittest.TestCase):

    def setUp(self):
        self.maxDiff = None

        self.sanitize_arr = [
            "having_IP_Address",
            "URL_Length",
            "Shortining_Service",
            "having_At_Symbol",
            "double_slash_redirecting",
            "Prefix_Suffix",
            "having_Sub_Domain",
            "domain_registeration_length",
            "Favicon",
            "port",
            "HTTPS_token",
            "Request_URL",
            "URL_of_Anchor",
            "Links_in_tags",
            "SFH",
            "Submitting_to_email",
            "Abnormal_URL",
            "Redirect",
            "on_mouseover",
            "RightClick",
            "popUpWidnow",
            "Iframe",
            "age_of_domain",
            "DNSRecord",
            "web_traffic",
            "Page_Rank",
            "Google_Index",
            "Links_pointing_to_page",
            "Statistical_report",
            "Result",
            "  485_5468a44  _44   4 ?  $@e3   *   C cc    c D  ",
        ]

        self.sample_col_names = [
            'first name',
            'First_Name',
            'first_name',
            'firstname',
            'surname',
            'Surname',
            'SurName',
            'namek',
            'rfid',
            'idea',
            'id',
            'Id',
            'iD',
            'ID',
            'customer_id',
            'customer id',
        ]

        self.df_change_dtypes = pd.DataFrame({
            'a': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20],
            'b': [1.5, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20],
            'c': [1, '2', 3, 4, 5, 6, '7', 8, 9, 10, '11', 12, 13, 14, 15, 16, 17, 18, 19, 20],
            'd': ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20'],
            'e': [1, 2, 3, 4, '?', '?', 7, 8, 9, 10, 11, '?', 13, 14, 15, 16, 17, '?', 19, 20],
            'f': ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't'],
            'g': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20],
        })

        self.df_outlier_regression = pd.DataFrame({
            'a': [1, 2, 9000, 4, 5, 6],
            'label': [42, 58, 62, 47, 13, 75],
        })
        self.df_outlier_classification = pd.DataFrame({
            'a': [5, 10, 7, 19, -99999, 17],
            'label': [1, 0, 0, 0, 1, 0],
        })

        self.df_fill_nan = pd.DataFrame({
            'a': [1, 2, 3, 4, 5, 6],
            'b': [np.nan, 2, 3, 4, np.nan, 6],
        })

        self.df_suggest_conversion = pd.DataFrame({
            'a': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
            'b': ['1', np.nan, '3', '4', '5', 6, '7', '8', 9, '10', '11'],
            'c': ['?', '2', 3, 4, 5, 6, 7, 8, 9, 10, 11],
        })

    def test_sanitize(self):
        new_cols = sanitize(self.sanitize_arr)

        self.assertListEqual(
            new_cols,
            [
                "having_ip_address",
                "url_length",
                "shortining_service",
                "having_at_symbol",
                "double_slash_redirecting",
                "prefix_suffix",
                "having_sub_domain",
                "domain_registeration_length",
                "favicon",
                "port",
                "https_token",
                "request_url",
                "url_of_anchor",
                "links_in_tags",
                "sfh",
                "submitting_to_email",
                "abnormal_url",
                "redirect",
                "on_mouseover",
                "right_click",
                "pop_up_widnow",
                "iframe",
                "age_of_domain",
                "dns_record",
                "web_traffic",
                "page_rank",
                "google_index",
                "links_pointing_to_page",
                "statistical_report",
                "result",
                "485_5468a44_44_4_e3_c_cc_c_d",
            ]
        )

    def test_change_dtypes(self):
        self.df_change_dtypes = change_dtypes(
            self.df_change_dtypes,
            {
                'a': 'float',
                'b': int,
                'c': 'int',
                'd': int,
                'e': float,
                'g': 'category',
            }
        )

        self.assertEqual(self.df_change_dtypes['a'].dtype, float)
        self.assertEqual(self.df_change_dtypes['b'].dtype, 'int')
        self.assertEqual(self.df_change_dtypes['c'].dtype, 'int')
        self.assertEqual(self.df_change_dtypes['d'].dtype, int)
        self.assertEqual(self.df_change_dtypes['e'].dtype, float)
        self.assertEqual(self.df_change_dtypes['g'].dtype, 'category')

    def test_remove_outliers(self):
        self.df_outlier_regression = remove_outliers(self.df_outlier_regression,
                                                     label_col='label',
                                                     std_coeff=1.5)
        self.df_outlier_classification = remove_outliers(self.df_outlier_classification,
                                                         label_col='label',
                                                         std_coeff=1.5)

        np.testing.assert_array_equal(list(self.df_outlier_regression['a']),
                                      [1.0, 2.0, np.nan, 4.0, 5.0, 6.0])
        np.testing.assert_array_equal(list(self.df_outlier_classification['a']),
                                      [5, 10, 7, 19, np.nan, 17])

    def test_fill_nan(self):
        self.df_fill_nan = fill_nan(self.df_fill_nan, 'mean')
        self.assertListEqual(list(self.df_fill_nan['a']),
                             [1, 2, 3, 4, 5, 6])
        self.assertListEqual(list(self.df_fill_nan['b']),
                             [3.75, 2, 3, 4, 3.75, 6])

        # the above test fills the null values in numeric columns of df
        # with mean. Therefore, reset the df back to original
        # to test for filling null values with median
        self.setUp()

        self.df_fill_nan = fill_nan(self.df_fill_nan, 'median')
        self.assertListEqual(list(self.df_fill_nan['a']),
                             [1, 2, 3, 4, 5, 6])
        self.assertListEqual(list(self.df_fill_nan['b']),
                             [3.5, 2, 3, 4, 3.5, 6])

        with self.assertRaises(ValueError):
            fill_nan(self.df_fill_nan, 'asdf')
            fill_nan(self.df_fill_nan, 5.0)

    def test_suggest_conversion_dict(self):
        suggested_conversion_dict = suggest_conversion_dict(
            self.df_suggest_conversion)
        self.assertEqual(suggested_conversion_dict, {'b': float, 'c': float})

    def test_suggest_col_drop(self):
        cols_to_drop = spot_irrelevant_columns(self.sample_col_names)

        self.assertListEqual(
            cols_to_drop,
            [
                'first name',
                'First_Name',
                'first_name',
                'firstname',
                'surname',
                'Surname',
                'SurName',
                'id',
                'Id',
                'iD',
                'ID',
                'customer_id',
                'customer id',
            ]
        )

        # if __name__ == "__main__":
        #     unittest.main()


tests/test_data_cleaner.py
#Version: 1

import unittest
import pandas as pd
import numpy as np
from cleaner.data_cleaner import sanitize_column_names, change_dtypes, remove_outliers, fill_nan, preprocess


class TestDataCleaner(unittest.TestCase):

    def setUp(self):
        self.df = pd.DataFrame({"A": [1, np.nan, 3, 4, 5, 50], "b": ["$ 50.0", "$2,000.00", -1.0, 0.0, np.nan, 6.0],
                                "  485 5468a44 44   4 ?  $@e3   *   C cc    c D  ": ["z", np.nan, "asd", "?wa\n kk  a", "3456", "$%^&*"]})

    @classmethod
    def tearDownClass(cls):
        print(cls.final_df.info())
        print(cls.final_df)

    def test_sanitize_column_names(self):
        new_df = sanitize_column_names(self.df)
        new_cols = list(new_df.columns)

        self.assertListEqual(
            new_cols, ["a", "b", "485_5468a44_44_4_e3_c_cc_c_d"])

    def test_change_dtypes(self):
        new_df = change_dtypes(self.df, {"b": float})

        self.assertEqual(new_df["b"].dtype, float)

    # def test_remove_outliers(self):
    #     new_df = remove_outliers(self.df, 1.5)

    def test_fill_nan(self):
        new_df = fill_nan(self.df, 'mean')
        self.assertEqual(list(new_df["A"]), [
                         1.0, 12.6, 3.0, 4.0, 5.0, 50.0])

        # the above test fills the null values in numeric columns of df
        # with mean. Therefore, reset the df back to original
        # to test for filling null values with median
        self.setUp()

        new_df = fill_nan(self.df, 'median')
        self.assertEqual(list(new_df["A"]), [
                         1.0, 4.0, 3.0, 4.0, 5.0, 50.0])

        with self.assertRaises(ValueError):
            fill_nan(self.df, 'asdf')
            fill_nan(self.df, 5.0)

    def test_preprocess(self):
        new_df = sanitize_column_names(self.df)
        new_df = preprocess(new_df, {"a": int, "b": float, "485_5468a44_44_4_e3_c_cc_c_d": 'category'},
                            1.5, 'median')

        # made it a class attribute inorder to use it in the
        # tearDownClass (to print the df and its info)
        self.__class__.final_df = new_df


if __name__ == "__main__":
    unittest.main()


dfcleaner/dflogger.py
import pandas as pd
import os


def change_logger(enable_logging, log_dir=''):
    def decorator(function):
        if not enable_logging:
            return function

        def wrapper(df, *args, **kwargs):
            old_df = df.copy(deep=True)
            df = function(df, *args, **kwargs)

            change_df = _df_diff(old_df, df)

            log_filepath = os.path.join(log_dir,
                                        '{}_log.csv'.format(function.__name__))

            change_df.to_csv(log_filepath)

            return df
        return wrapper
    return decorator


def _df_diff(oldFrame, newFrame):
    dfBool = (oldFrame != newFrame).stack()
    diff = pd.concat([oldFrame.stack()[dfBool],
                      newFrame.stack()[dfBool]], axis=1)
    diff.columns = ["old", "new"]
    return diff


dfcleaner/cleaner.py
import pandas as pd
import numpy as np
import re
from dfcleaner.dflogger import change_logger

ENABLE_LOGGING = False
LOG_DIR = '.'


@change_logger(ENABLE_LOGGING, LOG_DIR)
def preprocess(df, column_dtype_conversion_dictionary={}, std_coeff=1.5, fill_na_method='median', label_col=None):
    '''
    A convinient function that 
        - changes the datatypes of columns according to the 
            values given in the 'column_dtype_conversion_dictionary'
            argument
        - drops duplicate rows
        - if there is a target(label) column then drops all rows where that
            column value is null
        - removes outliers according to the std coefficient given as a parameter
            (doesn't consider the target(label) column to check for outliers)
        - fills nan values according to the fill_na_method parameter

    Note: This function performs all the above said actions in the 
        same order as mentioned.

    All the functions that are used here can also be used
    independently.

    Returns: 
        A cleaned up dataframe

    Args:
        df: pandas.DataFrame object

        column_dtype_conversion_dictionary: dictionary having keys as the 
            column name and value as the desired dtype

        std_coeff: coefficient of standard deviation in outlier removal

        fill_na_method: 'mean' or 'median'. nan values of each column will
            be replaced by that column's mean or median

        label_col: the target(label) column name (if any) as a string
    '''
    # df = sanitize_column_names(df)
    df = change_dtypes(df, column_dtype_conversion_dictionary)
    df = df.drop_duplicates()

    if label_col is not None:
        df = df.dropna(subset=[label_col])

    df = remove_outliers(df, std_coeff, label_col=label_col)
    df = fill_nan(df, fill_na_method, label_col=label_col)

    return df


def sanitize(arr):
    '''
    for each string in the array, this function will
        - keeps only alphanumeric, space and underscore characters
        - replaces multiple consecutive spaces with a single space
        - strips leading and trailing white spaces
        - replace spaces with underscores
        - convert CamelCase to snake_case
        - remove multiple consecutive underscores again
        - convert the whole string into lowercase

    Note: This function performs all the above said actions in the 
        same order as mentioned

    Returns: array of strings where the strings are 'sanitized'
    Args:
        arr: array of strings
    '''
    new_arr = []
    for string in arr:
        # only keep alphanumeric, space and underscore
        string = re.sub(r"[^A-Za-z0-9 _]", "", string)

        # remove multiple consecutive spaces
        string = re.sub(r" +", " ", string)

        # strip leading and trailing white spaces, lowercase
        # and replace space with underscore
        string = string.strip().replace(" ", "_")

        # convert CamelCase to snake_case
        string = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', string)
        string = re.sub('([a-z0-9])([A-Z])', r'\1_\2', string)

        # remove multiple consecutive underscores
        string = re.sub(r"_+", "_", string)

        # lower case
        string = string.lower()
        new_arr.append(string)

    return new_arr


def _filter_characters(dtype, element):
    '''
    if input element is not string then just returns the same element
    else,
    allows only digits and '.' and rejects other characters from
    given string

    Args:
        dtype: the desired dtype to convert the element into

    Example:
        input: "$ 5,000.00" (element is string therefore process it)
        output: 5000.00

        input: np.nan
        output: np.nan (returns the same element because element is not string)

        input: -6
        output: -6 (returns the same element because element is not string)

    '''
    if type(element) != str:
        return element
    else:
        filtered_characters = []
        for char in element:
            if char.isdigit() or char == '.':
                filtered_characters.append(char)

        if len(filtered_characters) == 0:
            return np.nan
        return dtype(''.join(filtered_characters))


@change_logger(ENABLE_LOGGING, LOG_DIR)
def change_dtypes(df, conversion_dictionary):
    '''
    This function will take a pandas.DataFrame and a 
    conversion dictionary as input and changes the datatypes 
    of the columns according to the conversion dictionary.

    Usage scenario 1:
        If a column where the values are like '$25.99', '$ 7.2', '$29,347.32' ...
        where you want to extract the float number and leave the string characters like '$', ',', ' ' ...
        you can just include the column name as key and float as value in the conversion dictionary
            Eg: {'col_name': float}

    Usage scenario 2:
        If a column where the values are actually float but because 
        of a single string character somewhere, the whole dtype 
        changes to Object
            Eg: a column has these values 27.6, 48.125, 24, ?, 74.32

        The ? is a string character in a numeric column trying to represent nan value
        In this case, simply include the key value pairs in the conversion dict 
        {'col_name': float}

    Returns: pandas.DataFrame with the changed datatypes

    Args:
        df: pandas.DataFrame object
        conversion_dict: dictionary with column names as keys and the 
            dtypes as values
            Eg: {'col_0': float, 'col_1': 'category', 'col_5': int}
    '''
    for col_name, dtype in conversion_dictionary.items():
        if dtype in [int, float]:
            if df[col_name].dtype not in [int, float]:
                df[col_name] = df[col_name].apply(
                    lambda x: _filter_characters(dtype, x))
            else:
                df[col_name] = df[col_name].astype(dtype)

        else:
            df[col_name] = df[col_name].astype(dtype)

    return df


@change_logger(ENABLE_LOGGING, LOG_DIR)
def remove_outliers(df, std_coeff=1.5, label_col=None):
    '''
    This function will take a dataframe and replaces all the outliers
    with np.nan.
    If target(label) column name is given, then it wont consider that
    column to check and remove outliers.

    The outliers are determined based on the std_coeff given as a parameter.

    Returns: pandas.DataFrame object without outlier values
    Args:
        df: pandas.DataFrame object
        std_coeff: the coefficient of standard deviation
            Eg: 1.5(recommended) or 3
        label_col: the target(label) column name (if any) as a string

    '''
    cols = list(df.columns)

    # consider only feat cols while removing outliers
    if label_col is not None:
        cols.remove(label_col)

    for col_name in cols:
        if df[col_name].dtype in [int, float]:
            df[col_name] = df[col_name].mask(df[col_name].sub(
                df[col_name].mean()).div(df[col_name].std()).abs().gt(std_coeff))

    return df


@change_logger(ENABLE_LOGGING, LOG_DIR)
def fill_nan(df, how, label_col=None):
    '''
    This function will take a pandas.DataFrame and fills all the 
    null values in all columns according to the method provided.

    If target(label) column name is given, then it wont consider that
    column to fill null values.

    Returns: pandas.DataFrame object without any null values

    Args:
        df: pandas.DataFrame
        how: 'median'(recommended) or 'mean'
        label_col: the target(label) column name (if any) as a string
    '''
    for col_name in df.columns:
        if df[col_name].dtype in [int, float]:
            if how == "median":
                df[col_name] = df[col_name].fillna(df[col_name].median())
            elif how == "mean":
                df[col_name] = df[col_name].fillna(df[col_name].mean())
            else:
                raise ValueError("'how' parameter must be 'mean' or 'median'")

    return df


def _can_convert_to_float(feat_col):
    '''
    determine if a feature column need to change their dtype from
    string to float by picking 10 random samples from that
    column and trying to parse it as float (float(value)).
    If the majority of them get converted to float then the
    dtype may be converted to float.
    eg: a numeric column might have a '?' instead of np.nan

    Args:
        feat_col: pandas.Series object (dataframe column)
    '''
    num_samples = 10
    required_sucess_ratio = 2/3
    samples = feat_col.sample(num_samples)

    successful_parse_count = 0
    for sample in samples:
        try:
            float(sample)
            successful_parse_count += 1
        except ValueError:
            pass

    if successful_parse_count > num_samples*required_sucess_ratio:
        return True
    return False


def suggest_conversion_dict(df):
    # use the _can_convert_to_float() function
    # if the column has nan then definitely int dtype
    '''
    This function checks if any string columns can be converted into float
    Eg:
        - a column may have numeric values but in string format 
            like, '12.0', '15.7'.
        - a column may have special symbols or characters instead 
            of np.nan, like, '?', 'null', 'na' along with normal 
            numbers like 12, 14.26, 85.15, ...
        - because of a single string value, the whole column might 
            have the 'Object' or 'category' datatype

    This function also checks if any numeric (int or float) columns can be converted into categorical columns with the dtype 'category'.
    Eg:
        - a column namely has_credit_card may have binary values 1 or 0.
        - This column can be considered as categorical because the 
            ratio of the number of unique values in the column to 
            the total number of values in the column is very small 
            (smaller than a threshold of 0.01)

    returns: dictionary that can be passed as an argument to the change_dtypes() function

    Args:
        df: pandas.DataFrame
    '''

    suggested_conversion_dict = {}
    for col in df.columns:
        if df[col].dtype in [int, float]:
            if _can_convert_to_category(df[col]):
                suggested_conversion_dict[col] = 'category'

        else:
            if _can_convert_to_float(df[col]):
                suggested_conversion_dict[col] = float

    return suggested_conversion_dict


def _can_convert_to_category(feat_col, threshold=0.01):
    '''
    determine if a numeric column is actually categorical by
    looking at what percentage of unique values are there when
    compared to the total number values
    Eg: 'has_credit_card' might have values 1 or 0
        which is technically numeric(int) but actually categorical
    '''
    uniq_len = len(feat_col.unique())
    total_len = len(feat_col)

    if uniq_len / total_len < threshold:
        return True

    return False


def spot_irrelevant_columns(cols):
    '''
    if there are any columns like names, first_names, ID
    then they are suggested to be dropped

    Args:
        cols: a list of column names

    returns:
        a list of columns to drop
    '''
    cols_to_drop = []
    for col in cols:
        if re.search(r'name\b', col.lower()):
            cols_to_drop.append(col)
        if re.search(r'(\b|[ _])id\b', col.lower()):
            cols_to_drop.append(col)

    return cols_to_drop
